core-backend/common/src/utils.rs

```diff
@@ -87,25 +58,55 @@ fn smart_truncate(s: &mut String, max_bytes: usize) {
     }
 }
 
-#[derive(Debug, Copy, Clone, derive_more::Display)]
-pub struct LimitedStr<'a>(&'a str);
+/// Wrapped string to fit `core-backend::TRIMMED_MAX_LEN` amount of bytes.
+///
+/// The `Cow` is used to avoid allocating a new `String` when the `LimitedStr` is
+/// created from a `&str`.
+///
+/// Plain `str` is not used because it can't be properly encoded/decoded via scale codec.
+#[derive(
+    TypeInfo, Encode, Decode, Debug, Clone, derive_more::Display, PartialEq, Eq, PartialOrd, Ord,
+)]
+#[codec(crate = scale)]
+pub struct LimitedStr<'a>(Cow<'a, str>);
+
 impl<'a> LimitedStr<'a> {
-    const INIT_ERROR_MSG: &'static str = concat!(
+    const INIT_ERROR_MSG: &str = concat!(
         "String must be less than ",
         stringify!(TRIMMED_MAX_LEN),
         " bytes."
     );
 
-    pub fn new(s: &'a str) -> Result<Self, &'static str> {
+    #[track_caller]
+    pub const fn from_small_str(s: &'a str) -> Self {
```

Comment:
If we really wanna use panic here, then use unwrap_or_else, as it is common way. Also, why do we need to specify str lifetime ?

Actually I'm not in glad with using panic here. Better to return an error, and make unwrap_or_else in places where we really wanna panic in wrong case. And you have already try_from

Reply:
We cannot use `try_from().expect("msg")` in constant context. The main point of `LimitedStr::from_small_str` is optimization for global names.
```rust
/// zero cost
const GLOBAL_NAMES: &[LimitedStr<'static>] = &[
    LimitedStr::from_small_str(GLOBAL_NAME_GAS),
];

/// `try_from` isn't const fn, `except(...)` isn't stabilized yet in const fns
const GLOBAL: LimitedStr<'static> = LimitedStr::try_from(GLOBAL_NAME_GAS).except("failed")
```

Reply:
Ok agree